incremental.go

```diff
@@ -0,0 +1,267 @@
+	// the push state and error (if occurred).
+	AfterPush func(*IncrementalProducerPush, error)
+}
+
+func (o *IncrementalProducerOptions) norm() {
+	if o.Format == nil {
+		o.Format = ProtobufFormat
+	}
+
+	if o.Compression == nil {
+		o.Compression = GZipCompression
+	}
+
+	if o.Interval <= 0 {
+		o.Interval = time.Minute
+	}
+}
+
+// IncrementalProducerPush contains the state of the last push.
+type IncrementalProducerPush struct {
+	// Producer exposes the current producer state.
+	*IncrementalProducer
+	// Updated indicates is the push resulted in an update.
+	Updated bool
+}
+
+// IncrementalProducer produces a continuous incremental feed.
+type IncrementalProducer struct {
+	bucket    bfs.Bucket
+	ownBucket bool
+	opt       IncrementalProducerOptions
+
+	ctx   context.Context
+	stop  context.CancelFunc
+	modfn LastModifiedFunc
+	pfn   IncrementalProduceFunc
+
+	numWritten, lastPush, lastMod int64
+}
+
+// NewIncrementalProducer inits a new incremental feed producer.
+func NewIncrementalProducer(ctx context.Context, bucketURL string, opt *IncrementalProducerOptions, modfn LastModifiedFunc, pfn IncrementalProduceFunc) (*IncrementalProducer, error) {
+	bucket, err := bfs.Connect(ctx, bucketURL)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := NewIncrementalProducerForBucket(ctx, bucket, opt, modfn, pfn)
+	if err != nil {
+		_ = bucket.Close()
+		return nil, err
+	}
+	p.ownBucket = true
+
+	return p, nil
+}
+
+// NewIncrmentalProducerForRemote starts a new incremental feed producer for a bucket.
+func NewIncrementalProducerForBucket(ctx context.Context, bucket bfs.Bucket, opt *IncrementalProducerOptions, modfn LastModifiedFunc, pfn IncrementalProduceFunc) (*IncrementalProducer, error) {
+	var o IncrementalProducerOptions
+	if opt != nil {
+		o = *opt
+	}
+	o.norm()
+
+	ctx, stop := context.WithCancel(ctx)
+	p := &IncrementalProducer{
+		bucket: bucket,
+		ctx:    ctx,
+		stop:   stop,
+		opt:    o,
+		modfn:  modfn,
+		pfn:    pfn,
+	}
+
+	// run initial push
+	if _, err := p.push(); err != nil {
+		_ = p.Close()
+		return nil, err
+	}
+
+	// start continuous loop
+	// use error group
+	go p.loop()
+
+	return p, nil
+}
+
+// Close stops the producer.
+func (p *IncrementalProducer) Close() error {
+	p.stop()
+	if p.ownBucket {
+		return p.bucket.Close()
+	}
+	return nil
+}
+
+// LastPush returns time of last push attempt.
+func (p *IncrementalProducer) LastPush() time.Time {
+	return timestamp(atomic.LoadInt64(&p.lastPush)).Time()
+}
+
+// LastModified returns time at which the remote feed was last modified.
+func (p *IncrementalProducer) LastModified() time.Time {
+	return timestamp(atomic.LoadInt64(&p.lastMod)).Time()
+}
+
+// NumWritten returns the number of values produced during the last push.
+func (p *IncrementalProducer) NumWritten() int {
+	return int(atomic.LoadInt64(&p.numWritten))
+}
+
+func (p *IncrementalProducer) loadManifest() (*Manifest, error) {
+	remote := bfs.NewObjectFromBucket(p.bucket, "manifest.json")
```

Comment:
you could keep the reference to the manifest open for the duration of the producer life-time, i.e. do this as part of `NewIncrmentalProducerForBucket` and close on `Close`